src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,316 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
+                        krb5_const_principal princ, const void *opts,
+                        const krb5_db_entry *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_san;
+    const struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    /* Verify the client SAN. */
+    ret = verify_client_san(context, req_opts->plgctx, req_opts->reqctx,
+                            req_opts->cb, req_opts->rock, princ, &valid_san);
+    if (ret == ENOENT) {
+        ret = KRB5_PLUGIN_NO_HANDLE;
+        goto out;
+    }
+
+    if (ret)
+        goto out;
+
+    if (!valid_san) {
+        pkiDebug("%s: did not find an acceptable SAN in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+    }
+
+out:
+    return ret;
+}
+static krb5_error_code
+pkinitsrv_eku_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                        const uint8_t *cert, size_t cert_len,
+                        krb5_const_principal princ, const void *opts,
+                        const krb5_db_entry *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_eku;
+    const struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    /* Verify the client EKU. */
+    ret = verify_client_eku(context, req_opts->plgctx, req_opts->reqctx,
+                            &valid_eku);
+    if (ret)
+        goto out;
+
+    if (!valid_eku) {
+        pkiDebug("%s: did not find an acceptable EKU in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_INCONSISTENT_KEY_PURPOSE;
+    }
+
+out:
+    return ret;
+}
+
+static krb5_error_code
+certauth_pkinitsrv_san_initvt(krb5_context context, int maj_ver, int min_ver,
+                              krb5_plugin_vtable vtable)
+{
+    krb5_certauth_vtable vt;
+
+    if (maj_ver != 1)
+        return KRB5_PLUGIN_VER_NOTSUPP;
+    vt = (krb5_certauth_vtable)vtable;
+    vt->name = "pkinitsrv";
+    vt->authorize = pkinitsrv_san_authorize;
+    vt->init = NULL;
+    vt->fini = NULL;
+    vt->free_ind = NULL;
+    return 0;
+}
+
+static krb5_error_code
+certauth_pkinitsrv_eku_initvt(krb5_context context, int maj_ver, int min_ver,
+                              krb5_plugin_vtable vtable)
+{
+    krb5_certauth_vtable vt;
+
+    if (maj_ver != 1)
+        return KRB5_PLUGIN_VER_NOTSUPP;
+    vt = (krb5_certauth_vtable)vtable;
+    vt->name = "pkinitsrv";
+    vt->authorize = pkinitsrv_eku_authorize;
+    vt->init = NULL;
+    vt->fini = NULL;
+    vt->free_ind = NULL;
+    return 0;
+}
+
+/*
+ * Do certificate auth based on match expressions in the pkinit_cert_match
+ * attribute string.  An expression should be in the same form as those used for
+ * the pkinit_cert_match configuration option.  Multiple expressions are
+ * separated by ':'.  The certificate is authorized if all expressions are
```

Comment:
This comment still needs to be updated now that the code supports only one expression.